# 01 Data Loading
**Agenda for this notebook:**

- Load the raw PakWheels dataset
- Inspect basic structure (shape, columns, data types)
- Understand the features available for car price prediction
- Document initial findings

**Project Notebooks Plan:**

1. `01_data_loading.ipynb` ← Current
2. `02_data_cleaning.ipynb`
3. `03_grouping_aggregation.ipynb`
4. `04_feature_engineering.ipynb`
5. `05_visualization.ipynb`

# Importing Required Libraries

- pandas is the main library for loading and working with tabular data
- We import it with the standard alias pd
- This is the first step in every data science notebook

In [1]:
import pandas as pd

# Loading the Raw Dataset

- We load the data from the raw/ folder
- Always use a variable for the file path (good professional habit)
- This step shows us how many car listings we have

In [4]:
# Load the raw PakWheels dataset
raw_data_path = '../data/raw/pakwheels.csv'   # Update filename if different
df = pd.read_csv(raw_data_path)

print(f"Shape: {df.shape[0]:,} rows and {df.shape[1]} columns")

Shape: 1,394 rows and 18 columns


# Checking Column Names

- We inspect the exact column names to understand what information is available
- This helps us plan which features are useful for predicting car price
- Common columns in such datasets: Make, Model, Year, Price, KM, etc.

In [5]:
# Display all column names
print("Column Names:")
for col in df.columns:
    print(f"- {col}")

Column Names:
- title
- url
- price
- year
- mileage
- engine
- transmission
- fuel_type
- city
- seller_type
- ad_id
- posted_date
- scraped_at
- source_type
- source_url
- make
- model_slug
- record_type


# First Look at the Data

- df.head() shows the first 5 rows
- This helps us understand what each column actually contains
- We can see real examples of car listings (price, year, mileage, etc.)

In [6]:
# Display first 5 rows
df.head()

,title,url,price,year,mileage,engine,transmission,fuel_type,city,seller_type,ad_id,posted_date,scraped_at,source_type,source_url,make,model_slug,record_type
0,Honda N Box 2022 Custom GL for Sale9.7/10,https://www.pakwheels.com/used-cars/honda-n-bo...,PKR 38.75 lacs,2022.0,"19,198 km",660 cc,Automatic,Petrol,Lahore,Individual,main_ad_10843516,Automatic,4/23/2026 0:00,used-car,https://www.pakwheels.com/used-cars/search/-/,NaN,NaN,NaN
1,Honda N Wgn 2023 L for Sale9.5/10,https://www.pakwheels.com/used-cars/honda-n-wg...,PKR 40.25 lacs,2023.0,"20,997 km",660 cc,Automatic,Petrol,Lahore,Individual,main_ad_10865190,Automatic,4/23/2026 0:00,used-car,https://www.pakwheels.com/used-cars/search/-/,NaN,NaN,NaN
2,Toyota Hilux 2004 Double Cab for Sale,https://www.pakwheels.com/used-cars/toyota-hil...,PKR 35 lacs,2004.0,"180,000 km",3000 cc,Manual,Diesel,Peshawar,Individual,main_ad_11347129,Manual,4/23/2026 0:00,used-car,https://www.pakwheels.com/used-cars/search/-/,NaN,NaN,NaN
3,Toyota Hilux 2017 Revo V Automatic 3.0 for Sale,https://www.pakwheels.com/used-cars/toyota-hil...,PKR 88.5 lacs,2017.0,"97,000 km",3000 cc,Automatic,Diesel,Peshawar,Individual,main_ad_11393981,Automatic,4/23/2026 0:00,used-car,https://www.pakwheels.com/used-cars/search/-/,NaN,NaN,NaN
4,Honda Civic 2019 Oriel 1.8 i-VTEC CVT for Sale,https://www.pakwheels.com/used-cars/honda-civi...,PKR 56.85 lacs,2019.0,"82,000 km",1800 cc,Automatic,Petrol,Rawalpindi,Individual,main_ad_11390502,Automatic,4/23/2026 0:00,used-car,https://www.pakwheels.com/used-cars/search/-/,NaN,NaN,NaN


# Better View of First Few Rows

- We use df.head(5).T to transpose the data
- This makes it easier to read all columns clearly for each car
- This view helps us understand the structure better

In [7]:
# Better view - transpose first 5 rows
df.head(5).T

,0,1,2,3,4
title,Honda N Box 2022 Custom GL for Sale9.7/10,Honda N Wgn 2023 L for Sale9.5/10,Toyota Hilux 2004 Double Cab for Sale,Toyota Hilux 2017 Revo V Automatic 3.0 for Sale,Honda Civic 2019 Oriel 1.8 i-VTEC CVT for Sale
url,https://www.pakwheels.com/used-cars/honda-n-bo...,https://www.pakwheels.com/used-cars/honda-n-wg...,https://www.pakwheels.com/used-cars/toyota-hil...,https://www.pakwheels.com/used-cars/toyota-hil...,https://www.pakwheels.com/used-cars/honda-civi...
price,PKR 38.75 lacs,PKR 40.25 lacs,PKR 35 lacs,PKR 88.5 lacs,PKR 56.85 lacs
year,2022.0,2023.0,2004.0,2017.0,2019.0
mileage,"19,198 km","20,997 km","180,000 km","97,000 km","82,000 km"
engine,660 cc,660 cc,3000 cc,3000 cc,1800 cc
transmission,Automatic,Automatic,Manual,Automatic,Automatic
fuel_type,Petrol,Petrol,Diesel,Diesel,Petrol
city,Lahore,Lahore,Peshawar,Peshawar,Rawalpindi
seller_type,Individual,Individual,Individual,Individual,Individual


# Basic Information About Dataset

- df.info() shows data types and missing values
- This helps us identify which columns need cleaning
- We expect many object (text) columns that will need conversion

In [8]:
# Get basic information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1394 entries, 0 to 1393
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         1331 non-null   object 
 1   url           1331 non-null   object 
 2   price         1331 non-null   object 
 3   year          1331 non-null   float64
 4   mileage       1331 non-null   object 
 5   engine        605 non-null    object 
 6   transmission  605 non-null    object 
 7   fuel_type     1331 non-null   object 
 8   city          1331 non-null   object 
 9   seller_type   1331 non-null   object 
 10  ad_id         1331 non-null   object 
 11  posted_date   1331 non-null   object 
 12  scraped_at    1331 non-null   object 
 13  source_type   1331 non-null   object 
 14  source_url    1331 non-null   object 
 15  make          0 non-null      float64
 16  model_slug    0 non-null      float64
 17  record_type   0 non-null      float64
dtypes: float64(4), object(14)
me

# Checking Missing Values

- df.isnull().sum() shows exactly how many missing values are in each column
- This is crucial for planning our data cleaning strategy
- We expect high missing values in make, model_slug, engine, etc.

In [9]:
# Count missing values in each column
missing_values = df.isnull().sum()
print(missing_values)

title             63
url               63
price             63
year              63
mileage           63
engine           789
transmission     789
fuel_type         63
city              63
seller_type       63
ad_id             63
posted_date       63
scraped_at        63
source_type       63
source_url        63
make            1394
model_slug      1394
record_type     1394
dtype: int64


# Statistical Summary of Numerical Columns

- df.describe() gives us count, mean, min, max for numerical columns
- This helps us understand the range of year and spot any unrealistic values early
- We will focus on year for now as it is the only useful numeric column currently

In [10]:
# Get statistical summary
df.describe()

,year,make,model_slug,record_type
count,1331.000000,0.0,0.0,0.0
mean,2018.510894,NaN,NaN,NaN
std,7.246601,NaN,NaN,NaN
min,1978.000000,NaN,NaN,NaN
25%,2016.000000,NaN,NaN,NaN
50%,2021.000000,NaN,NaN,NaN
75%,2024.000000,NaN,NaN,NaN
max,2026.000000,NaN,NaN,NaN


# Data Loading Summary

- We have successfully loaded and inspected the raw dataset
- Key findings: 1,394 total rows, many missing values, price and mileage stored as text
- This completes the initial exploration of the PakWheels dataset

In [11]:
# Final summary of data loading
print("=== 01_data_loading.ipynb Summary ===")
print(f"Total rows loaded     : {len(df):,}")
print(f"Rows with price data  : {df['price'].notnull().sum():,}")
print(f"Year range            : {df['year'].min()} - {df['year'].max()}")
print(f"Columns with missing values: {df.isnull().sum()[df.isnull().sum() > 0].count()}")
print("\n Data loading completed!")

=== 01_data_loading.ipynb Summary ===
Total rows loaded     : 1,394
Rows with price data  : 1,331
Year range            : 1978.0 - 2026.0
Columns with missing values: 18

 Data loading completed!
